packages/playwright-core/src/server/injected/domUtils.ts

```diff
@@ -72,6 +72,12 @@ export function isElementVisible(element: Element): boolean {
     }
     return false;
   }
+  // Element.checkVisibility checks for content-visibility and also looks at
+  // styles up the flat tree including user-agent ShadowRoots, such as the
+  // details element for example.
+  // @ts-ignore Typescript doesn't know that checkVisibility exists yet.
+  if (Element.prototype.checkVisibility && !element.checkVisibility())
```

Comment:
@josepharhar Sorry for the long delay!

I looked at the implementation and I think this is not 100% compatible, but pretty close to it.
- Browser looks at `opacity: 0`, while playwright does not. Can we pass `ignoreOpacity` option?
- Browser checks some kind of "inert" criteria - what is that?
- Browser checks that `visibility === visible`, while Playwright only checks for `hidden`. I think browser's behavior is actually better, but I'd like to make a separate change for that.

Other comments:
- We should probably move line 63 after this check, to not be redundant, wdyt?
- What about display:contents elements? Are there any plans to address those?

Reply:
> Browser looks at opacity: 0, while playwright does not. Can we pass ignoreOpacity option?

> Browser checks that visibility === visible, while Playwright only checks for hidden. I think browser's behavior is actually better, but I'd like to make a separate change for that.

Opacity and visibility are not looked at by default: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/element.idl;l=29-30;drc=8b23a425b4a5e75caa5b91c78b062a38b1f92aea

I'll add the parameters and set them to false just to make it clearer though.

> Browser checks some kind of "inert" criteria - what is that?

The first iteration of checkVisibility had inert, but I actually removed it. It shouldn't exist anywhere.

> We should probably move line 63 after this check, to not be redundant, wdyt?

Done

> What about display:contents elements? Are there any plans to address those?

Hmm... I suppose that if an element it display:contents, then it shouldn't have a LayoutObject, which would make checkVisibility return false. The code earlier in this function already checks for display:contents though and won't continue to checkVisibility.